homeassistant/components/tplink/__init__.py

```diff
@@ -0,0 +1,102 @@
+"""Component to embed TP-Link smart home devices."""
+import logging
+
+from homeassistant.const import CONF_HOST
+from homeassistant import config_entries
+from homeassistant.helpers import config_entry_flow
+import homeassistant.helpers.config_validation as cv
+import voluptuous as vol
+
+_LOGGER = logging.getLogger(__name__)
+
+TPLINK_HOST_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string
+})
+
+TPLINK_SCHEMA = vol.Schema({
+    vol.Optional('light'): vol.All(cv.ensure_list, [TPLINK_HOST_SCHEMA]),
+    vol.Optional('switch'): vol.All(cv.ensure_list, [TPLINK_HOST_SCHEMA]),
+    vol.Optional('discovery', default=True): bool,
+})
+
+DOMAIN = 'tplink'
+REQUIREMENTS = ['pyHS100==0.3.3']
+
+
+async def _async_has_devices(hass):
+    """Return if there are devices that can be discovered."""
+    from pyHS100 import Discover
+
+    def discover():
+        devs = Discover.discover()
+        return devs
+    return await hass.async_add_executor_job(discover)
+
+
+async def async_setup(hass, config):
+    """Set up the TP-Link component."""
+    conf = config.get(DOMAIN)
+
+    if conf is not None:
+        hass.data[DOMAIN] = TPLINK_SCHEMA(conf)
+    else:
+        hass.data[DOMAIN] = {'light': [], 'switch': [], 'discovery': True}
+
+    if conf is not None:
+        hass.async_create_task(hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': config_entries.SOURCE_IMPORT}))
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up TPLink from a config entry."""
+    from pyHS100 import SmartBulb, SmartPlug, SmartDeviceException
+
+    devices = dict()
+
+    if hass.data[DOMAIN]["discovery"]:
+        devs = await _async_has_devices(hass)
+        _LOGGER.info("Discovered %s TP-Link smart home devices", len(devs))
+        devices.update(devs)
+
+    for type_ in ['light', 'switch']:
+        for entry in hass.data[DOMAIN][type_]:
+            try:
+                host = entry["host"]
+                if type == 'light':
+                    dev = SmartBulb(host)
+                elif type_ == 'switch':
+                    dev = SmartPlug
+                devices[host] = dev
+                _LOGGER.info("Succesfully added %s %s: %s",
+                             type_, host, dev)
+            except SmartDeviceException as ex:
+                _LOGGER.error("Unable to initialize %s %s: %s",
+                              type_, host, ex)
+
+    for dev in devices.values():
+        if isinstance(dev, SmartPlug):
```

Comment:
why not merge this with previous for loop ?

Reply:
This is done separately to allow overriding detected types and to report back on unsupported devices. I'm not very satisfied with the structure, so I'll think something to improve this.

Reply:
The reporting can happen if the backend library supports (and returns) a non-supported devices as a part of the discovery process, making it at least visible when the component needs to be updated.

The override happens if a device is discovered, but also listed manually. Would it be better to raise an exception here for such misconfiguration? Alternatively I could get rid of separate setting arrays for lights and switches, and detect them during the runtime, avoiding this two-loop setup.

Reply:
let's keep it for now. People really like their manual config